EIPS/eip-6662.md

```diff
@@ -0,0 +1,116 @@
+category: ERC
+created: 2023-03-09
+requires: 712, 4337, 4804
+---
+
+## Abstract
+
+This ERC proposes a new **IAccountMetadata** interface as an extension for [ERC-4337](./eip-4337.md) to store authentication data on-chain to support a more user-friendly authentication model.
+
+## Motivation
+
+In this proposal, we propose a new **IAccountMetadata** interface as an extension for ERC-4337 **IAccount** interface. With this new interface, users can store authentication data on-chain through one-time publishing, allowing dApps to proactively fetch it from the chain to support a more flexible and user-friendly authentication model. This will serve as an alternative to the current authentication model where users need to log in with a wallet every time and push account-related information to dApps by connecting the wallet in advance.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Authentication Flow
+
+![Authentication Flow](../assets/eip-6662/auth-flow.png)
+
+In the new authentication workflow, users use AA compatible smart contract accounts as their wallet addresses. **Authenticator** could be anything but holding the private key to sign users' operations. For example, it can be an offline authenticator mobile app or an online cloud service. **Relay** is an online service responsible for forwarding requests from dApps to the Authenticator. If the authenticator is online, it can play the role of Relay service and listen to dApps directly.
+
+### Interface
+
+To support the new authentication workflow, this ERC proposes a new **IAccountMetadata** interface as an extension of **IAccount** interface defined by ERC-4337.
+
+```
+interface IAccountMetadata {
+  struct AuthenticatorInfo {
+    // a list of service URIs to relay message from dApps to authenticators
+    string[] relayURI;
+    // a JSON string or URI pointing to a JSON file describing the
+    // schema of AuthenticationRequest. The URI should follow ERC-4804
+    // if the schema file is stored on-chain
+    string schema;
+  }
+
+  function getAuthenticationInfo() external view returns(AuthenticatorInfo[] memory);
+}
+```
+
+The relay endpoint should accept an AuthenticationRequest object as input. The format of the AuthenticationRequest object is defined by the schema field at AuthenticationInfo.
+
+Following is a schema example which supports end to end encryption, where we pack all encrypted fields into an encryptedData field. The symbol **“\$e2ee”** indicates that the field is encrypted and the encrypted value is stored at the encryptedData field. Here we only list three basic fields but there may be more fields per schema definition.
+
+```
+{
+    "title": "AuthenticationRequest",
+    "type": "object",
+    "properties": {
+        "entrypoint": {
+            "type": "string" | "$e2ee",
```

Comment:
I am unsure if `$e2ee` is a valid JSON schema type, could you clarify? 
https://json-schema.org/understanding-json-schema/reference/type.html

Reply:
it's a special prefix we used to indicate if the field is e2e encrypted. The type should be string. I updated the field to type to string and explained the symbol usage in the article.